### ISSUE TITLE
Allow html conditional comments and doctype

### ISSUE THREAD

**Issue by natew** [facebook/react#1035]
I'm hacking on an fully server and client compatible app using React. One obstactle I'm running into is it seems react doesn't like exclamation marks in your jsx-style code.

Here's an example of a pretty standard boilerplate for HTML:

```
<!doctype html>
  <!--[if IE 7]>    <html class="no-js ie ie7 ltie8 ltie9"><![endif]-->
  <!--[if IE 8]>    <html class="no-js ie ie8 ltie9"><![endif]-->
  <!--[if gt IE 8]> <html class="no-js ie gtie8"><![endif]-->
  <!--[if !IE]><!--><html class="no-js" xmlns:og="http://opengraphprotocol.org/schema/" xmlns:fb="http://www.facebook.com/2008/fbml" ><!--<![endif]-->
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <title></title>
      <meta name="description" content="">
      <meta name="viewport" content="width=device-width, initial-scale=1">

      <link rel="stylesheet" href="/css/app.css" type="text/css" media="all" />
    </head>
    <body>
      <!--[if lt IE 8]>
        <p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</p>
      <![endif]-->

      {this.props.page}

      <!-- <script src="/js/vendor.js"></script> -->
      <script src="/js/app.js"></script>
    </body>
  </html>
```

I think it's getting stuck on the conditional comments in this example.

I would prefer to keep this code in the HTML style, and to render it from the server without making major changes, it needs to be inside a react class. I'd be open to trying to hack this together, I may take a stab sometime this coming week.

**Comment by adjavaherian** [comment]
Should I not be using React to render complete documents before serving?

**Comment by zackify** [comment]
@adjavaherian You shouldn't need to.

**Comment by JulianLeviston** [comment]
Actually I think React CAN mount higher than body client-side: In the caes where you want react to render an iFrame... What's the current "best of breed" for iFrame generation, where you want react to control the iFrame's content? This seems largely impossible at the moment.

Also, does dangerouslySetInnerHTML cause the browser to (re-)run JS that is in that string? 

Also, does React do its diffing on that dangerous setting? Doing a full replace would cause the entire document of the iFrame to stutter on update, and all the resources in it to reload on most browsers if it's a set rather than a diff/update, I think?

Not sure how the diffing would work with sanboxed JS running in the iFrame though?

Also, does React have a grouping component that doesn't actually do anything but mount other components? (kind of a silent component). I couldn't seem to see one. Probably should learn more about the internal architecture, I guess.

**Comment by gaearon** [comment]
>Also, does React have a grouping component that doesn't actually do anything but mount other components? (kind of a silent component). I couldn't seem to see one. Probably should learn more about the internal architecture, I guess.

I don't know much about it but I saw these links:

https://developer.zendesk.com/blog/rendering-to-iframes-in-react
http://jsfiddle.net/YXgzf/

**Comment by JulianLeviston** [comment]
Thanks, @gaearon :) I found NodeList and glossary subsequent to my last comment. And yeah, I've read the zendesk article... I've also implemented his stuff to a degree, but it has large holes in it when it comes to stuff above body (like styles, script, etc.), and the whole approach feels like a big hack-around... (fighting the code) hence this thread :)

**Comment by JulianLeviston** [comment]
Is there any movement on this, @wyuenho @petehunt ?

**Comment by ftorre104** [comment]
We render full react pages, using react-router, on the server.

An easy workaround was just adding the doctype to the returned markup on the server:
`res.send('<!DOCTYPE html>' + markup);`

Initially I thought it might cause an invariant on the client, but it doesn't seem to cause any problems.
To be sure, you could prepend the same doctype string in your client-side entry/rehydrate file.

**Comment by xcatliu** [comment]
I've found a hack way to insert comments to html:
https://nemisj.com/conditional-ie-comments-in-react-js/

**Comment by nuweb** [comment]
+1 doctype for fully-page server rendering.

**Comment by vizath** [comment]
For my usecase, I use React to render emails using `React.renderToStaticMarkup()`.
It would be nice to add conditionnal comments for Outlook.

**Comment by zackify** [comment]
In React 0.14, rendering the `<body>` tag (at least with the test utils) raises an invariant violation. Not sure if I should be mentioning that here, or make a new issue. I figured it applies to this.

**Comment by zpao** [comment]
@zackify That's unrelated, please file a new issue with steps to repro

**Comment by schickling** [comment]
Any update on this?

**Comment by zpao** [comment]
No update.

**Comment by baygeldin** [comment]
I've just ran into this. The ability to render full page with just react can help to deal with SEO, title and many other things with much flexibility. And you won't even need something like helmet or react-document-title. And it simply seems logical to me to handle whole application state within react. While prepending doctype is easy, other things are really hard or impossible \*sigh\*.

**Comment by baygeldin** [comment]
Well, I found out that some people say that changing head with virtual DOM can push some browsers to go crazy, which is definitely worse than not having whole app within react. Maybe, it wasn't a good idea after all...

**Comment by jimfb** [comment]
@baygeldin Specifically, can you define "go crazy"?  Which browsers, changing which nodes?

**Comment by baygeldin** [comment]
@jimfb I guess I should have said "an unexpected behavior". I couldn't find a good example, but from what I understood the main arguments are: 1) Browsers do not expect replacing the whole nodes in the head, so their behavior is not standardized 2) It's a common practice for browser-extensions to inject external resources to the head, react will override those if it handles the head tag. Also, according to MDN, document.head is read-only. Sorry, I don't have concrete examples and time to investigate it :( I decided to go the standard way of doing things.

**Comment by syrnick** [comment]
+1 for this. It's quite a bit unexpected that server-side react can't actually render the whole page.

**Comment by optimalisatie** [comment]
Hi!

I have found a new method to include HTML comments (e.g. conditional IE comments) in JSX and React components using a Web Component.

https://github.com/optimalisatie/react-jsx-html-comments

```jsx
<react-comment>[if lt IE 8]</react-comment>
<p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="http://browsehappy.com/">upgrade your browser</a> to improve your experience.</p>
<react-comment>[endif]</react-comment>
```

**Comment by sumbad** [comment]
Hello, I added additional component Head in my server-side to resolve this situation:

```
class Head extends React.Component{
    render() {
        let dangerousInnerHTML: string[] = this.props.children.map(value => {
            if (typeof value !== 'string')
                return ReactDOMServer.renderToStaticMarkup(value);
            else
                return value;
        });

        return (
            <head dangerouslySetInnerHTML={{ __html: dangerousInnerHTML.join('') }}>
            </head>
        );
    }
}

```

and now I can just use it:

```
        return (
            <html lang="ru">
                <Head>
                    <meta charset="utf-8" />
                    <title>Example</title>
                    <meta name="viewport" content="width=device-width, initial-scale=1" />
                    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
                    <link rel="stylesheet" href="stylesheets/views/index/index.css" />
                    <script src="javascript/index/bundle.js"></script>
                    {`
                        <!--[if lte IE 8]><link rel="stylesheet" href="javascript/css/ie8.css" /><![endif]-->
                        <!--[if lte IE 9]><link rel="stylesheet" href="javascript/css/ie9.css" /><![endif]-->
                    `}
                 </Head>
...
```

Maybe this is not perfect solution and requires some additional code. Please, review and say what do you think about it. Thanks.

**Comment by PsiXdev** [comment]
@sumbad The children can contain an array, this requires a separate check, because ReactDOMServer.renderToStaticMarkup can not render arrays.

**Comment by dominikwilkowski** [comment]
This is definitely something needed for server side rendering. For right now we have to inject it after we call `renderToStaticMarkup`. An official way would be much better.

**Comment by khilnani** [comment]
+1

**Comment by gaearon** [comment]
Closing as there is an easy workaround (string manipulation before sending the markup for doctype, and `dangerouslySetInnerHTML` for comments). React also doesn’t support IE8 anymore, and the only remaining browser for which conditional comments are relevant is IE9.

Adding support for comments or doctype would have to span multiple projects (React, Babel, ESLint, etc) and yields little benefit, so it is unlikely we’ll be working on this.

**Comment by dominikwilkowski** [comment]
For everyone still finding this issue and wondering just how to use `dangerouslySetInnerHTML` in your code, have a look at this article: https://nemisj.com/conditional-ie-comments-in-react-js/

TL;DR:
```jsx
renderHead() {
  return (
    var comment = '<!--[if lte IE 9]><script src="/public/media.match.js"></script><![endif]-->';
    <head>
      <title>Website title</title>
      <meta name="react-comment-hack" 
          dangerouslySetInnerHTML={{__html: comment}}>
      </meta>
    </head>
  );
}
```

**Comment by shaodahong** [comment]
still error 
```
Invariant Violation: meta is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`
```

**Comment by ebrahimiaval** [comment]
i'm use this to fix it
` template = '<!DOCTYPE html>' + ReactDOMServer.renderToString(<Template />)`

**Comment by loganvolkers** [comment]
@sumbad stumbled across this and years later it looks like you're on the right path.

Your solution didn't work with Context, so here's an updated solution that does that and adds some generic helpers in the same approach as `styled-components`.

```js
export function RawHTMLWrapper({ as, children = [], contexts = [] }) {
  const values = contexts.map((c) => useContext(c));

  const toHtml = (value) => {
    if (typeof value !== "string") {
      const element = contexts.reduce((el, c, idx) => {
        return <c.Provider value={values[idx]}>{el}</c.Provider>;
      }, value);
      return ReactDOMServer.renderToStaticMarkup(element);
    } else {
      return value;
    }
  };

  let dangerousInnerHTML;
  if (Array.isArray(children)) {
    dangerousInnerHTML = children.map(toHtml).join("");
  } else {
    dangerousInnerHTML = toHtml(children);
  }

  const props = {
    dangerouslySetInnerHTML: { __html: dangerousInnerHTML }
  };
  return React.createElement(as, props);
}

const elements = [
  "div",
  "tr",
  "table",
  "td",
  "h1",
  "h2",
  "h3",
  "h4",
  "h5",
  "h6"
];
const reducer = (acc, el) => {
  const Componet = ({ children, contexts }) => (
    <RawHTMLWrapper as={el} contexts={contexts}>
      {children}
    </RawHTMLWrapper>
  );
  return { ...acc, [el]: Componet };
};

export const raw = elements.reduce(reducer, {});
```

And then using it via `raw.div` etc.

```js
const FooContext = createContext("foo");

export default function App() {
  return (
    <FooContext.Provider value="definitely not foo">
      <raw.div contexts={[FooContext]}>
        <raw.h1>
          <span>Hello CodeSandbox</span>
        </raw.h1>
        <h2>Start editing to see some magic happen!</h2>
        {`<h3>And some static HTML</h3>`}
        {`<!-- and some comments -->`}
        <raw.h4 contexts={[FooContext]}>
          Uses context: <Consumes />
        </raw.h4>
        {`<h3>And more static HTML</h3>`}
      </raw.div>
    </FooContext.Provider>
  );
}

function Consumes() {
  const foo = useContext(FooContext);

  return <span>{foo}</span>;
}
```


You can see it running here: https://codesandbox.io/s/react-raw-html-wrapper-hc2jq?file=/src/RawHTMLWrapper.js:91-674

**Comment by Ciannode-deasync** [comment]
@natew https://stylesheets.space/questions/ie9-fails-to-display-libefore-element-correctly
